### PR TITLE
Check for generic structs in "needs marshalling" check

### LINKED ISSUE #90828: Blittability check doesn't account for generic structs
[MarshallingRequired](https://github.com/dotnet/runtime/blob/76d454907d00722609fe170d7d97bfbe644abc50/src/coreclr/vm/dllimport.cpp#L3176) check does not have a case for generic value types and will report them as requiring marshalling at runtime and throw an invalid program exception. 

A minimal repro is below:

```C#
public static unsafe class Program
{
    public static unsafe void Main()
    {
        unsafe {
            void* m = (void*)(delegate* unmanaged[MemberFunction]<void*, MyStruct<nint>, int, int> )&ABI_Method;
            var i = ((delegate* unmanaged[MemberFunction]<MyStruct<int>, int, int>)m)(new MyStruct<int>(), 1);
        }
    }
    [global::System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute(CallConvs = new[] { typeof(global::System.Runtime.CompilerServices.CallConvMemberFunction) })]
    internal static int ABI_Method(void* __this_native, MyStruct<nint> __p_native, int size)
    { return 0;}

}

struct MyStruct<T> where T: unmanaged {}
```

### PR DESCRIPTION
**Pull request by jtschuster**
Fixes https://github.com/dotnet/runtime/issues/90828.

### CODE REVIEW

**Review by jkoritzinsky — COMMENTED** [review]
Can we add a negative test that throws InvalidProgramException with a generic class type?

**Comment by jtschuster** [comment]
> Can we add a negative test that throws InvalidProgramException with a generic class type?

I don't think we can, the analyzer will create an error for classes and structs with NonBlittable generic parameters.

**Review by jkoritzinsky — APPROVED** [review]

File: src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs [hunk 1]
```diff
@@ -225,9 +225,9 @@ public static void TestUnmanagedCallersOnlyWithGeneric()
 
         Assert.Equal(0, ((delegate* unmanaged<MaybeBlittable<nint>, int>)&MaybeBlittableGenericStruct)(new MaybeBlittable<nint>()));
 
-        Assert.Throws<InvalidProgramException> (() => ((delegate* unmanaged<NotBlittable<int>, int>)&InvalidGenericUnmanagedCallersOnlyParameters.GenericClass)(new NotBlittable<int>()));
+        Assert.Throws<MarshalDirectiveException> (() => ((delegate* unmanaged<NotBlittable<int>, int>)&InvalidGenericUnmanagedCallersOnlyParameters.GenericClass)(new NotBlittable<int>()));
```
Comment (jkoritzinsky) [inline src/tests/Interop/UnmanagedCallersOnly/UnmanagedCallersOnlyTest.cs:228]: Can we change this to use the "wrong" type on the managed side (like `nint`) to ensure that we get to the UnmanagedCallersOnly callee? The `MarshalDirectiveException` is on the managed->unmanaged side.

File: src/coreclr/vm/dllimport.cpp [hunk 2]
```diff
@@ -3316,8 +3316,11 @@ BOOL NDirect::MarshalingRequired(
                 FALLTHROUGH;
 
             case ELEMENT_TYPE_VALUETYPE:
+            case ELEMENT_TYPE_GENERICINST:
             {
                 TypeHandle hndArgType = arg.GetTypeHandleThrowing(pModule, &emptyTypeContext);
+                if(!hndArgType.IsValueType())
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/dllimport.cpp:3322]: Perhaps there is something above, but this doesn't seem fully sufficient. What about is a value type with a managed field?

File: src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il [hunk 3]
```diff
@@ -240,4 +240,76 @@
...
+        int32 GenericStructWithObjectField (
+            valuetype InvalidCSharp.MaybeBlittable`1<object> param
+        ) cil managed
+    {
+        .custom instance void [System.Runtime.InteropServices]System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute::.ctor() = (
+            01 00 00 00
+        )
+        // Method begins at RVA 0x24c4
+        // Header size: 1
+        // Code size: 2 (0x2)
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il:309]:
```suggestion
```

File: src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il [hunk 4]
```diff
@@ -240,4 +240,76 @@
...
+        int32 GenericClass (
+            class InvalidCSharp.NotBlittable`1<int32> param
+        ) cil managed
+    {
+        .custom instance void [System.Runtime.InteropServices]System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute::.ctor() = (
+            01 00 00 00
+        )
+        // Method begins at RVA 0x24c4
+        // Header size: 1
+        // Code size: 2 (0x2)
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il:292]:
```suggestion
```

File: src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il [hunk 5]
```diff
@@ -240,4 +240,76 @@
...
+	extends [System.Runtime]System.Object
+{
+	// Fields
+	.field private !T Value
+    // Methods
+    .method public hidebysig specialname rtspecialname
+        instance void .ctor () cil managed
+    {
+        // Method begins at RVA 0x20a0
+        // Code size 7 (0x7)
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/UnmanagedCallersOnly/InvalidCallbacks.il:262]:
```suggestion
```

File: src/coreclr/vm/dllimport.cpp (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/dllimport.cpp:3322]: Ah. The logic is below for this case.

File: src/coreclr/vm/dllimport.cpp [hunk 6]
```diff
@@ -3319,7 +3319,8 @@ BOOL NDirect::MarshalingRequired(
             case ELEMENT_TYPE_GENERICINST:
             {
                 TypeHandle hndArgType = arg.GetTypeHandleThrowing(pModule, &emptyTypeContext);
-                if(!hndArgType.IsValueType())
+                bool isValidGeneric = !IsValidForGenericMarshalling( hndArgType.AsMethodTable(), false, runtimeMarshallingEnabled));
```
Comment (jkoritzinsky) [inline src/coreclr/vm/dllimport.cpp:3322]:
```suggestion
                bool isValidGeneric = !IsValidForGenericMarshalling( hndArgType.GetMethodTable(), false, runtimeMarshallingEnabled));
```

File: src/coreclr/vm/dllimport.cpp [hunk 7]
```diff
@@ -3319,7 +3319,8 @@ BOOL NDirect::MarshalingRequired(
...
+                if(!hndArgType.IsValueType() ||  isValidGeneric)
```
Comment (jkoritzinsky) [inline src/coreclr/vm/dllimport.cpp:3323]:
```suggestion
                bool isValidGeneric = IsValidForGenericMarshalling( hndArgType.GetMethodTable(), false, runtimeMarshallingEnabled));
                if(!hndArgType.IsValueType() ||  !isValidGeneric)
```

**Review by jkoritzinsky — APPROVED** [review]

**Pull request merged**